### PR TITLE
Enumerate subtypes by approximating type skolems to their upper bound

### LINKED ISSUE #9657: Incorrect 'warning: unreachable code' on Type Member
Given:

```scala
scala> sealed trait PowerSource 
defined trait PowerSource

scala> case object Petrol extends PowerSource
defined object Petrol

scala> case object Pedal extends PowerSource
defined object Pedal

scala> sealed abstract class Vehicle {
     |   type A <: PowerSource
     | }
defined class Vehicle

scala> case object Bicycle extends Vehicle {
     |   type A = Pedal.type
     | }

scala> case class Bus(fuel: Int) extends Vehicle {
     |   type A = Petrol.type
     | }
defined class Bus

scala> case class Car(fuel: Int) extends Vehicle {
     |   type A = Petrol.type
     | }
```

And now the `refuel` function. Observe that there's an `unreachable` code warning that's not true.

```scala
scala> def refuel[P <: Petrol.type](vehicle: Vehicle {type A = P} ): Vehicle = vehicle match {
     |   case Car(_) => Car(100)
     |   case Bus(_) => Bus(100)
     | }
<console>:20: warning: unreachable code
         case Bus(_) => Bus(100)
                           ^
refuel: [P <: Petrol.type](vehicle: Vehicle{type A = P})Vehicle

scala> refuel(Car(100))
res6: Vehicle = Car(100)

scala> refuel(Bus(5))
res7: Vehicle = Bus(100)
```

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes scala/bug#9657

### CODE REVIEW

**Comment by dwijnand** [comment]
Very similar to #9472.

**Comment by SethTisue** [comment]
Related to #9472, yes, but that one fixes a pretty easy-to-understand bug, whereas here, the example code in scala/bug#9657 seems rather esoteric and has a lot of moving parts. So it's hard for me to convince myself that this is both correct and complete, but on the other hand, it's esoteric enough that I'm inclined to shrug if all the tests still pass.

I guess my main review question: how sure are we that the original bug report is minimized? If a simpler test case can be constructed, I think that would be valuable. If not, oh well.

**Comment by dwijnand** [comment]
The fishy bit, IMO, is the indirection of `P <: Petrol.type` - there's no reason for that not to be inlined in the parameter type.  So instead of can make a `PetroleumBasedPowerSource` sealed trait, with Petrol and Coal as case object subclasses, make PetroleumBasedPowerSource the upper bound of the type parameter and double check again.

**Review by lrytz — COMMENTED** [review]
I think this looks good.

Do you have an explanation why adding a `case Bicycle` is not marked as unreachable? `enumerateSubtypes` seems to correctly return `Bus(_), Car(_)`.

**Comment by dwijnand** [comment]
> Do you have an explanation why adding a `case Bicycle` is not marked as unreachable? `enumerateSubtypes` seems to correctly return `Bus(_), Car(_)`.

I hadn't yet studied the reachability logic that layers on top of the rest of the patmat/SAT logic, so I went ahead and did that.  I'm going to record my notes, as it's been taxing enough for one day.  Here's where things go wonky:

First of all, there's a disagreement on what the variable can be:

```
reachability, vars:
V1: Vehicle{type A = P} ::= Bus | Car | null// Set(Bus, Bicycle, null, Car) // = x1
```

The domain knows it can't be Bicycle, but the variables constants still contains the Bicycle constant.

When it's looping case-by-case to see if they're reachable, it starts correctly:

```
trying to reach:
  And(Set(Not(V1=null#1), V1=Bus#3))
under prefix:
  And(Set(Or(Set(V1=Bus#3, V1=Car#2, V1=null#1)), AtMostOne(List(V1=Bus#3, V1=Car#2, V1=null#1))))
  Or(Set(V1=null#1, Not(V1=Car#2)))
reached model:
  V1=Car#2 -> false
  V1=null#1 -> false
  V1=Bus#3 -> true
var assignments:
V1(=x1: Vehicle{type A = P}) == (Bus)  != (Car, null)
```

English: if V1 must be Bus, Car xor null, and it's not Car, can V1 be a Bus?  Yes, if V1 is Bus, that's satisfied.

Then a wild Bicycle reappears:

```
trying to reach:
  V1=Bicycle#4
under prefix:
  And(Set(Or(Set(V1=Bus#3, V1=Car#2, V1=null#1)), AtMostOne(List(V1=Bus#3, V1=Car#2, V1=null#1))))
  Or(Set(V1=null#1, Not(V1=Car#2)))
  Or(Set(V1=null#1, Not(V1=Bus#3)))
reached model:
  V1=Car#2 -> false
  V1=Bus#3 -> false
  V1=null#1 -> true
  V1=Bicycle#4 -> true
var assignments:
V1(=x1: Vehicle{type A = P}) == (null, Bicycle)  != (Car, Bus)
```

English: if V1 (same as before) must be Bus, Car xor null, and it's not Car AAAAND it's not Bus, can it be Bicycle?  Yes, if V1 is both the null and non-null Bicycle at the same time... 🙄 

```
        // should never be more than one value in trues...
```

**Comment by dwijnand** [comment]
> First of all, there's a disagreement on what the variable can be:
> 
> ```
> reachability, vars:
> V1: Vehicle{type A = P} ::= Bus | Car | null// Set(Bus, Bicycle, null, Car) // = x1
> ```
> 
> The domain knows it can't be Bicycle, but the variables constants still contains the Bicycle constant.

This comes from registering the variable = constant equalities from the case defs, i.e. the actual `case Bicycle`:

```scala
      object gatherEqualities extends PropTraverser {
        override def apply(p: Prop) = p match {
          case Eq(v, NullConst) if !modelNull => vars += v // not modeling equality to null
          case Eq(v, c)                       => vars += v; v.registerEquality(c)
          case _                              => super.apply(p)
        }
      }
```

And the reason before this patch we got:

```
test/files/neg/t9657.scala:14: warning: unreachable code
    case Bus(_)  => Bus(100) // was: "unreachable code" warning 1
                       ^
```

is because there the SAT solving fails instantly:

```
reachability, vars: V1: Vehicle{type A = P} ::= null// Set(Bicycle, Car, Bus, null) // = x1
equality axioms:
  And(Set(Or(Set(V1=null#1))))
trying to reach:
  And(Set(Not(V1=null#1), V1=Bus#3))
under prefix:
  V1=null#1
  Or(Set(V1=null#1, Not(V1=Car#2)))
```

If V1 is null, can it be (non-null) Bus?  No, unreachable.

Where does `V1=null` come from?  Well it's `Or(Set(V1=null#1, Not(V1=Car#2)))` except here without the Car part, because it doesn't consider Car to be a subtype of `Vehicle { type A = P }` with the `P <: Petrol.type` skolem.

I guess at least I feel less concerned about the domain being a subset and not equal to the variable consts.  But something is going wrong as the solver is assigning V1 to a constant it can't be.

**Comment by dwijnand** [comment]
> English: if V1 (same as before) must be Bus, Car xor null, and it's not Car AAAAND it's not Bus, can it be Bicycle? Yes, if V1 is both the null and non-null Bicycle at the same time... 🙄
> 
> ```
>         // should never be more than one value in trues...
> ```

Ah, I see.  It's not obvious from the debug text, but it's not Bicycle the type, it's Bicycle the value.  So it's `null`, within the domain of Vehicle, and `Bicycle` as in the ValueConst Bicycle.

Tough...  I think I'll have to park this detail if/until I get smarter and/or more knowledgeable.

**Comment by lrytz** [comment]
Thanks! It sounds like with enough time you can work your way through; in any case the fix in this PR seems right, and the wild Bicycle appearing is a separate issue.

**Review by lrytz — APPROVED** [review]

**Comment by dwijnand** [comment]
> there's no reason for that not to be inlined in the parameter type. So instead of can make a `PetroleumBasedPowerSource` sealed trait, with Petrol and Coal as case object subclasses, make PetroleumBasedPowerSource the upper bound of the type parameter and double check again.

I've changed my patch to hand this (via deSkolemize), but I'm not sure about it...

**Comment by dwijnand** [comment]
Have a look...  The test case comes from chatting about Seth about whether the original test case is minimised enough and/or makes sense.  I wrote what I felt was a slightly more reasonable reason to run into the original problem ("less fishy") and it ended up opening a can of worms.

There's a pun, for @som-snytt, the pun-maestro: What came first?  The fishy code or the can of worms?

File: test/files/pos/t9657-less-fishy.scala [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+case object Coal   extends NonRenewablePowerSource
+case object Pedal  extends PowerSource
+
+sealed abstract class Vehicle { type A <: PowerSource }
+case object Bicycle       extends Vehicle { type A = Pedal.type  }
+case class Bus(fuel: Int) extends Vehicle { type A = Petrol.type }
+case class Car(fuel: Int) extends Vehicle { type A = Petrol.type }
+
+object Test {
+  def refuel[P <: NonRenewablePowerSource](vehicle: Vehicle { type A = P }): Vehicle = vehicle match {
```
Comment (dwijnand) [inline test/files/pos/t9657-less-fishy.scala]: Chatting with Fengyun, perhaps this is still fishy... why not just:

```scala
def refuel2(vehicle: Vehicle { type A <: NonRenewablePowerSource }): Vehicle = ...
```

?

Comment (dwijnand) [reply test/files/pos/t9657-less-fishy.scala]: On this basis, I dropped this test case and, thus, the less-sure patch I had pushed.

**Comment by joroKr21** [comment]
This still fails: https://scastie.scala-lang.org/WgY7KC87QeeGv9RPbA29Yw (reported on the shapeless Discord channel)

**Comment by joroKr21** [comment]
I tried changing it to `approximateAbstracts` instead, because the upper bound makes sense only in covariant positions.
But then the warning flipped from unreachable code to non-exhaustive match 🤔

**Pull request merged**